File: rmi-chat/src/server/ChatService.java
package server;

import java.rmi.RemoteException;
import java.rmi.registry.Registry;
import java.rmi.server.UnicastRemoteObject;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import common.IChat;
import common.IChatCallback;

public class ChatService implements IChat {

	private Map<String, IChatCallback> users = new HashMap<String, IChatCallback>();

	public ChatService(Registry registry, String name) throws RemoteException {
		registry.rebind(name, UnicastRemoteObject.exportObject(this, 0));
	}

	@Override
	public void login(String name, IChatCallback callback) throws RemoteException {
		if (users.containsKey(name)) {
			callback.onDeny();
		} else {
			final List<String> names = new ArrayList<String>();
			names.add("*");
			users.entrySet().stream().forEach(entry -> {
				try {
					entry.getValue().onAddUser(name);
					names.add(entry.getKey());
				} catch (RemoteException e) {
				}
			});
			users.put(name, callback);
			callback.onAccept(names.toArray(new String[names.size()]));
		}
	}

	@Override
	public void logout(IChatCallback callback) throws RemoteException {
		final Optional<String> name = getName(callback);
		if (name.isPresent()) {
			users.remove(name.get());
			users.entrySet().forEach(entry -> {
				try {
					entry.getValue().onRemoveUser(name.get());
				} catch (RemoteException e) {
				}
			});
			callback.onExit();
		}
	}

	@Override
	public void send(String to, String message, IChatCallback callback) throws RemoteException {
		final Optional<String> name = getName(callback);
		if (name.isPresent()) {
			users.entrySet()
				.stream()
				.filter(entry -> "*".equals(to) 
						|| entry.getKey().equals(to)
						|| entry.getValue().equals(callback))
				.forEach(entry -> {
					try {
						entry.getValue().onReceive(name.get(), message);
					} catch (RemoteException e) {
					}
				});
		}
	}
	
	private Optional<String> getName(IChatCallback callback) {
		return users.entrySet()
				.stream()
				.filter(entry -> entry.getValue().equals(callback))
				.map(entry -> entry.getKey())
				.findAny();
	}

}

File: WcfChat/Server/ChatService.cs
using System.ServiceModel;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using Common;

namespace Server
{
	internal class ChatService : IChat
	{
		private static IDictionary<string, IChatCallback> users = new ConcurrentDictionary<string, IChatCallback>();
		public void Login(string name)
		{
			if (users.ContainsKey(name))
			{
				Callback.OnDeny();
			}
			else
			{
				var names = new List<string>();
				names.Add("*");
				for (var enumeration = users.GetEnumerator(); enumeration.MoveNext();)
				{
					enumeration.Current.Value.OnAddUser(name);
					names.Add(enumeration.Current.Key);
				}
				users[name] = Callback;
				Callback.OnAccept(names.ToArray());
			}
		}

		public void Logout()
		{
			var (name, callback) = users.AsEnumerable()
				.Where(e => e.Value == Callback)
				.Select(entry => (entry.Key, entry.Value))
				.First();
			users.Remove(name);
			users.AsParallel()
				.ForAll(entry => entry.Value.OnRemoveUser(name));
			callback.OnExit();
		}

		public void Send(string to, string message)
		{
			var from = users.AsEnumerable()
				.Where(entry => entry.Value == Callback)
				.Select(entry => entry.Key)
				.First();
			users.AsParallel()
				.Where(entry => to == "*" || entry.Key == to || Callback == entry.Value)
				.ForAll(entry => entry.Value.OnReceive(from, message));
		}

		private IChatCallback Callback => OperationContext.Current.GetCallbackChannel<IChatCallback>();
	}
}
